nautobot/core/management/commands/pre_migrate.py

```diff
@@ -120,6 +120,48 @@ def check_configcontext_uniqueness():
         )
 
 
+def check_interface_ipaddress_vrf_uniqueness(command):
+    """
+    Check for uniqueness enforcement changes for Interface and VMInterface IPAddress assignments.
+
+    - All IP Addresses assigned to a specific Interface or VMInterface must share the same VRF or have no VRF.
+    """
+
+    failed = False
+
+    command.stdout.write(command.style.WARNING(">>> Running interface/vminterface vrf uniqueness checks..."))
+
+    for ip_address_assignment in (
+        IPAddress.objects.values("assigned_object_type", "assigned_object_id")
```

Comment:
This "feels" backwards to me, but maybe I'm not understanding it. 
- Why are we iterating over all IPAddresses instead of iterating over all Interfaces and all VMInterfaces?
- Shouldn't we filter out / discard IPs that have no `assigned_object` at all (or (VM)Interfaces that have no `ip_addresses`) for efficiency?

Reply:
This query is providing all of the information that is required for this check directly from the database:

- Aggregation of `assigned_object_type`/`assigned_object_id`
- Count of unique vrfs associated to that `assigned_object_type`/`assigned_object_id` combination
- Count of `IPAddress` objects associated to that `assigned_object_type`/`assigned_object_id` combination with a null vrf (workaround for `null != null` and `COUNT` only returning non-null values)

Using this query:

```sql
SELECT 
  "ipam_ipaddress"."assigned_object_type_id", 
  "ipam_ipaddress"."assigned_object_id", 
  COUNT(
    DISTINCT "ipam_ipaddress"."vrf_id"
  ) AS "vrf_count", 
  COUNT("ipam_ipaddress"."id") FILTER (
    WHERE 
      "ipam_ipaddress"."vrf_id" IS NULL
  ) AS "vrf_null_count" 
FROM 
  "ipam_ipaddress" 
GROUP BY 
  "ipam_ipaddress"."assigned_object_type_id", 
  "ipam_ipaddress"."assigned_object_id";

```

We could probably put a `.filter(assigned_object_id__isnull=False, assigned_object_type_id__isnull=False)` in there to speed it up:

```sql
SELECT 
  "ipam_ipaddress"."assigned_object_type_id", 
  "ipam_ipaddress"."assigned_object_id", 
  COUNT(
    DISTINCT "ipam_ipaddress"."vrf_id"
  ) AS "vrf_count", 
  COUNT("ipam_ipaddress"."id") FILTER (
    WHERE 
      "ipam_ipaddress"."vrf_id" IS NULL
  ) AS "vrf_null_count" 
FROM 
  "ipam_ipaddress" 
WHERE 
  (
    "ipam_ipaddress"."assigned_object_id" IS NOT NULL 
    AND "ipam_ipaddress"."assigned_object_type_id" IS NOT NULL
  ) 
GROUP BY 
  "ipam_ipaddress"."assigned_object_type_id", 
  "ipam_ipaddress"."assigned_object_id";

```